### DISCUSSION TITLE
Photodiode square spontaneously black for short time

### DISCUSSION THREAD

**Discussion by petersaj** [bonsai-rx/bonsai#1434] [General]
Hi - I've included a photodiode square into my workflows to turn on/off with stimulus presentation, it works almost all of the time, but recently I have been seeing some examples where the photodiode registers a flip to black even though Bonsai shouldn't have flipped. 

Here's an example, where the blue trace is the photodiode readout, the red are timestamps where Bonsai flipped the photodiode, but there's a drop in the photodiode that wasn't intended by Bonsai at the green arrow:
![image](https://github.com/bonsai-rx/bonsai/assets/17985056/db7018c6-77ac-4ac8-8f7f-a6b416780e6e)

I'm pretty sure it's not an issue with the photodiode, because there's a decay that looks like the pixels turning to black, and the noise level at black is the same in that blip as when the photodiode is intentionally black. 

Here's my protocol (added .txt in filename for upload), the photodiode flip command is in DisplayStim > DrawGratingSequence > PhotodiodeFlip:
[lcr_passive.bonsai.txt](https://github.com/bonsai-rx/bonsai/files/11885072/lcr_passive.bonsai.txt)

I've also seen this same thing (again rarely) in a sparse noise workflow, which updates the photodiode square in a similar way.

Is there a reason that the photodiode square might flip to black unintentionally, or something I can check? I haven't managed to see it happen in real time yet since it's very rare, and I would think if it's something like a lagging computer issue that the frame draw might be delayed but the square wouldn't turn black.

**Comment by bruno-f-cruz** [comment]
Heya.
First of all, nicely organized workflow. Super readable!
I can't run it with a photodiode but looking at the workflow, I wonder if you can look at:

1. The layer of your quad. Right now both your `Quad` and `Stim` seem to be draw at the same depth (0). I wonder if your stim parameters are such that you, sometimes, draw the stimulus over the sync quad. Try to set the layer of the quad higher than everything else.
2.  Trying to flip the level logic of the synchronization. Basically invert the diode to be off during stimulation. I know this seems stupid, but I had a monitor a few years ago that had some sort of hardware problem and would just randomly flicker. The outcome was something similar to what you are seeing. Since the flicker was due to the backlight turning off, if you invert the logic, you should not see it during the stimulus and only during the ITI. That would at least allow you to figure out if it is a problem with your workflow or with the hardware.
3. Related to 2. Is there anything funky going on around the times of this toggle? For instance, if you look at the RealElapsedTime field of `RenderFrame`, is it the case that frames "choke" around that point in time?


I will keep thinking about other potential sources that might explain the problem. Meanwhile, keep me posted, I am curious about this one 
B

**Reply by petersaj** [reply]
Thanks, and thanks for the quick help! I tried changing the layers, but then think I figured out the source of the problem: we've got two monitors on the computer running bonsai, one for the stimulus and one outside the rig to show a different program. Some combination of clicking on various windows to select them makes the bonsai stimulus window flick to black very briefly. I have windows effects turned off, so it's something about window switching rather than a feature, but either way it looks like the error's on our side.

**Reply by petersaj** [reply]
So the issue in the end appears to be a Windows thing: when CreateWindow is in the fullscreen state, clicking into or out of the window flicks the window to black briefly. A way to get around this is to create a normal window with hidden window border and defined width/height, which produces ~the same display (actually not exactly, sometimes pixels hang off the screen even though the size matches the screen resolution) but gets rid of the flicker on click.
